[llvm-dwarfdump] Make --verify for .debug_names multithreaded.

ayermolo opened the pull request:
This PR makes verification of .debug_names acceleration table multithreaded. In local testing it improves verification of clang .debug_names from four minutes to under a minute.
This PR relies on a current mechanism of extracting DIEs into a vector. 
Future improvements can include creating API to extract one DIE at a time, or grouping Entires into buckets by CUs and extracting before parallel step.

Single Thread
4:12.37 real,   246.88 user,    3.54 sys,       0 amem,10232004 mmem
Multi Thread
0:49.40 real,   612.84 user,    515.73 sys,     0 amem, 11226292 mmem

Commit:
[llvm-dwarf] Make --verify for .debug_names multithreaded.

This PR makes verification of .debug_names acceleration table multithreaded.
In local testing it improves verification of clang .debug_names
from four minutes to under a minute.
This PR relies on a current mechanism of extracting DIEs into a vector.
Future improvements can include creating API to extract DIE at a time,
or grouping Entires into buckets by CUs and extracting before parallel
step.

Single Thread
4:12.37 real,   246.88 user,    3.54 sys,       0 amem,
10232004 mmem
0:49.40 real,   612.84 user,    515.73 sys,     0 amem, 11226292 mmem

adrian-prantl commented:
Does it make the output nondeterministic, or does it remain sorted?

youngd007 commented on llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:
```diff
@@ -32,7 +33,9 @@ struct DWARFSection;
 
 class OutputCategoryAggregator {
 private:
+  std::mutex WriteMutex;
   std::map<std::string, unsigned> Aggregation;
+  unsigned NumErrors = 0;
   bool IncludeDetail;
```
@ayermolo can you make sure you have the latest.  I actually had just modified the output aggregator to start tracking sub-categories for the JSON.  See https://github.com/llvm/llvm-project/pull/125062/files
There also were some tests added for JSON that would be good to run.  They _might_ be running on the build jobs assuming it would merge up to latest.  Less clear how github works 🤷‍♂️

youngd007 commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -284,6 +286,9 @@ static cl::opt<bool>
                 cat(DwarfDumpCategory));
 static opt<bool> Verify("verify", desc("Verify the DWARF debug info."),
                         cat(DwarfDumpCategory));
+static opt<unsigned> VerifyNumThreads(
+    "verify-num-threads", init(hardware_concurrency().compute_thread_count()),
+    desc("Number of threads to use for --verify."), cat(DwarfDumpCategory));
```
Do the threads only impact verifying debug names?  Should that be made clear for end users?  If it does have an effect, even better :D

ayermolo replied:
Yes it is optional. Right now it's only for debug-names, but as other parts of verification become parallel it will apply to them also. 
It defaults to max threads hardware supports.

youngd007 commented on llvm/test/tools/llvm-dwarfdump/verify_split_dwarf_debug_names_ftus.test:
```diff
@@ -5,7 +5,7 @@
 # RUN: yaml2obj %p/Inputs/verify_split_dwarf_debug_names_ftus_dwo.yaml > %t1/main.dwo
 # RUN: cd %t1
-# RUN: llvm-dwarfdump --verify %t1/a.out | FileCheck %s
+# RUN: llvm-dwarfdump --verify --verify-num-threads 1 %t1/a.out | FileCheck %s
```
It's good that all these tests show that the output with the change and a single thread is the same, but could we add a test where there are two threads?
As Adrian inquired, is the output deterministic?  I assume JSON definitely still is, but what about the raw output?  @adrian-prantl , is that what you're asking about?

ayermolo commented:
There are two components to report:
Example:
```
ErrorCategory.Report("Name Index references non-existing CU", [&]() {
          error() << formatv(
              "Name Index @ {0:x} references a non-existing CU @ {1:x}\n",
              NI.getUnitOffset(), Offset);
        });
```

Category that we output to report with counter. In example it's:
`Name Index references non-existing CU.`
And an error message that gets printed by way of lambda out that can change.  In this case based on the NameIndex Offset and CU offset.

```
formatv( "Name Index @ {0:x} references a non-existing CU @ {1:x}\n",
              NI.getUnitOffset(), Offset)
```

Raw output is not deterministic because it is printed as errors encountered. 
If we want deterministic output then we would need to store all the error messages, and sort them somehow.
If there are a lot of messages then this will blow up the memory.

ayermolo commented:
To add.
As a followup we can refactor:

```
void OutputCategoryAggregator::Report(
    StringRef s, std::function<void(void)> detailCallback) {
  std::lock_guard<std::mutex> Lock(WriteMutex);
  ++NumErrors;
  Aggregation[std::string(s)]++;
  if (IncludeDetail)
    detailCallback();
}
```
So that category is an enum.
We can then turn Aggregation into a vector, of vectors.
Where second vector will be some generic/compact data structure, that callback would know how to set.
We can then print all the categories after everything was processed. 
Just an idea.

clayborg commented on llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:
```diff
@@ -32,7 +33,9 @@ struct DWARFSection;
 private:
+  std::mutex WriteMutex;
   std::map<std::string, unsigned> Aggregation;
+  unsigned NumErrors = 0;
```
I would use `uint64_t` here instead of `unsigned`.

clayborg commented on llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:
```diff
@@ -42,6 +45,8 @@ class OutputCategoryAggregator {
   void Report(StringRef s, std::function<void()> detailCallback);
   void EnumerateResults(std::function<void(StringRef, unsigned)> handleCounts);
+  /// Return the number of errors that have been reported.
+  unsigned GetNumErrors() const { return NumErrors; }
```
uint64_t

clayborg commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -903,6 +908,7 @@ int main(int argc, char **argv) {
 
   bool Success = true;
   if (Verify) {
+    parallel::strategy = hardware_concurrency(VerifyNumThreads);
```
Where is this used?

clayborg commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -775,7 +780,7 @@ static bool handleBuffer(StringRef Filename, MemoryBufferRef Buffer,
       std::unique_ptr<DWARFContext> DICtx = DWARFContext::create(
           *Obj, DWARFContext::ProcessDebugRelocations::Process, nullptr, "",
-          RecoverableErrorHandler);
+          RecoverableErrorHandler, WithColor::defaultWarningHandler, true);
```
change `true` to `/*ThreadSafe=*/true`

youngd007 commented:
@ayermolo I downloaded this patch to try on a large binary and ended up with a seg fault.  I'll reach out internally to make sure this is resolved prior to landing.
Sadly, cannot request changes here

Commit:
addressed comments, fixed tsan issues

Commit:
fixed performance regression, removed locks, refactored how it checks if die should be in NameIndex. For latter it was hitting pathological case with bolted binaries. Making it 7x slower.

Commit:
fixed handling of corrupt name table, and handling errors in
verifyDebugNamesCULists.
Latter had a small bug that used return. This meant depending on which
NI was processed first, it might not get to all the errors.

felipepiovezan commented:
> Unfortunately adding lock_guard to DWARFVerifier::getNames kills all the performance. :( it invokes DIE.getShortName() which does recursive traversal of DIEs. This can include references to other CUs, thinlto case, or type units. Which triggers extracting all the DIEs for that TU/CU. Will need to think how to refactor this.

Didn't we have a "peek DIE name" function to address this issue?

Commit:
minor nits, made a test single threaded. I think technically speaking it's not legal

Commit:
removed check for DWARF5, it will pre-load DWARF4 also.

ayermolo commented:
> > Unfortunately adding lock_guard to DWARFVerifier::getNames kills all the performance. :( it invokes DIE.getShortName() which does recursive traversal of DIEs. This can include references to other CUs, thinlto case, or type units. Which triggers extracting all the DIEs for that TU/CU. Will need to think how to refactor this.
> 
> Didn't we have a "peek DIE name" function to address this issue?

I believe it's in only LLDB.
In it's implementation of DWARFUnit in Plugins/SymbolFile/DWARF

clayborg reviewed (commented):
Might be good to have to specify using multiple threads when verifying. So maybe opt into multi-threaded output to maintain deterministic output by default and allowing users to opt into non-deterministic output for speed? Something like
```
--verify-multithreaded
```
If this is specified, then we can take full advantage of the speed at the cost of non-deterministic output.

ayermolo commented:
> Might be good to have to specify using multiple threads when verifying. So maybe opt into multi-threaded output to maintain deterministic output by default and allowing users to opt into non-deterministic output for speed? Something like
> 
> ```
> --verify-multithreaded
> ```
> 
> If this is specified, then we can take full advantage of the speed at the cost of non-deterministic output.

I added a flag:
--verify-num-threads.
It defaults to system number of threads.
Instead set it to 1?

Jlalond commented on llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h:
```diff
@@ -793,6 +794,14 @@ class DWARFDebugNames : public DWARFAcceleratorTable {
+      return CurrentIndex->getNameTableEntry(idx + 1);
+    }
+
+    difference_type operator-(const NameIterator &other) const {
```
Can we get some comments on context on the new operators? I don't see where they are used elsewhere in this patch.

ayermolo replied:
Updated.
They are used in the parallelForEach implementation.
parallelFor(0, End - Begin, [&](size_t I) { Fn(Begin[I]); });

Commit:
updated with comment where new operators are used

clayborg commented:
> > Might be good to have to specify using multiple threads when verifying. So maybe opt into multi-threaded output to maintain deterministic output by default and allowing users to opt into non-deterministic output for speed? Something like
> > ```
> > --verify-multithreaded
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If this is specified, then we can take full advantage of the speed at the cost of non-deterministic output.
> 
> I added a flag: --verify-num-threads. It defaults to system number of threads. Instead set it to 1?

That will be better to avoid anyone having different or unexpected output. If we change the default it can affect users that rely on this.

clayborg commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -284,6 +286,9 @@ static cl::opt<bool>
                         cat(DwarfDumpCategory));
+static opt<unsigned> VerifyNumThreads(
+    "verify-num-threads", init(hardware_concurrency().compute_thread_count()),
+    desc("Number of threads to use for --verify."), cat(DwarfDumpCategory));
```
Probably best to expand this documentation. 
```
sc("Number of threads to use for --verify. Single threaded verification is the default unless this option is specified. This can cause the output to be non determinisitic, but can speed up verification and is useful when running with the summary only or JSON summary modes.")
```

Commit:
changed to default 1 thread, updated tests

Commit:
added option for max threads
